File: mcp-server/golang/mcp_server.go
package main

import (
	"bytes"
	"context"
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/exec"
	"os/signal"
	"path/filepath"
	"strings"
	"syscall"
	"time"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

// Implementation metadata for the MCP server
var impl = &mcp.Implementation{Name: "vscode-file-finder-go", Version: "0.1.0"}

// SearchFilesParams defines inputs for the search_files tool
// jsonschema tags are used by the SDK to derive the input schema
// keeping names aligned with the Python server version.
type SearchFilesParams struct {
	Name      string `json:"name" jsonschema:"Glob or pattern for file names"`
	Content   string `json:"content" jsonschema:"Substring / text to search inside files"`
	Directory string `json:"directory" jsonschema:"Root directory to start search (default: ".")"`
}

// OpenFileParams defines inputs for the open_file tool
type OpenFileParams struct {
	Path    string `json:"path" jsonschema:"Path to file or directory"`
	OpenDir bool   `json:"open_dir" jsonschema:"Treat path as directory"`
}

// resolve helper binary path: VS_CODE_HELPER_BIN or ./vscode-helper or LookPath("vscode-helper")
func helperBin() (string, error) {
	if env := strings.TrimSpace(os.Getenv("VS_CODE_HELPER_BIN")); env != "" {
		return env, nil
	}
	// Prefer local project binary like Python server
	local := "./vscode-helper"
	if st, err := os.Stat(local); err == nil && !st.IsDir() {
		return local, nil
	}
	if lp, err := exec.LookPath("vscode-helper"); err == nil {
		return lp, nil
	}
	return "", fmt.Errorf("vscode-helper binary not found; build it with: go build -o vscode-helper")
}

func runHelper(ctx context.Context, args ...string) (string, error) {
	bin, err := helperBin()
	if err != nil {
		return "", err
	}
	cmd := exec.CommandContext(ctx, bin, args...)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		msg := strings.TrimSpace(stderr.String())
		if msg == "" {
			msg = strings.TrimSpace(stdout.String())
		}
		if msg == "" {
			msg = err.Error()
		}
		return "", fmt.Errorf(msg)
	}
	out := strings.TrimSpace(stdout.String())
	return out, nil
}

// searchFiles implements the ToolHandlerFor signature by delegating to the helper binary.
func searchFiles(ctx context.Context, ss *mcp.ServerSession, params *mcp.CallToolParamsFor[SearchFilesParams]) (*mcp.CallToolResultFor[any], error) {
	p := params.Arguments
	var args []string
	args = append(args, "search")
	if strings.TrimSpace(p.Name) != "" {
		args = append(args, "--name", p.Name)
	}
	if strings.TrimSpace(p.Content) != "" {
		args = append(args, "--content", p.Content)
	}
	if dir := strings.TrimSpace(p.Directory); dir != "" && dir != "." {
		args = append(args, "--dir", dir)
	}
	out, err := runHelper(ctx, args...)
	if err != nil {
		return textResult("Error searching: " + err.Error()), nil
	}
	if out == "" {
		out = "(no matches)"
	}
	return textResult(out), nil
}

// openFile delegates to helper binary 'open' command exactly like Python server
func openFile(ctx context.Context, ss *mcp.ServerSession, params *mcp.CallToolParamsFor[OpenFileParams]) (*mcp.CallToolResultFor[any], error) {
	p := params.Arguments
	if strings.TrimSpace(p.Path) == "" {
		return textResult("Error: 'path' is required"), nil
	}
	var args []string
	args = append(args, "open")
	if p.OpenDir {
		args = append(args, "--dir")
	}
	// Pass the path as provided; the helper will resolve/validate and call 'code'
	args = append(args, p.Path)
	out, err := runHelper(ctx, args...)
	if err != nil {
		return textResult("Error opening: " + err.Error()), nil
	}
	if out == "" {
		// CLI prints confirmation; but ensure some response
		abs, _ := filepath.Abs(p.Path)
		out = "Opened: " + abs
	}
	return textResult(out), nil
}

func textResult(s string) *mcp.CallToolResultFor[any] {
	return &mcp.CallToolResultFor[any]{
		Content: []mcp.Content{&mcp.TextContent{Text: s}},
	}
}

// createServer constructs the MCP server and registers tools.
func createServer() *mcp.Server {
	server := mcp.NewServer(impl, nil)
	mcp.AddTool(server, &mcp.Tool{Name: "search_files", Description: "Search files by name and/or content starting at a directory."}, searchFiles)
	mcp.AddTool(server, &mcp.Tool{Name: "open_file", Description: "Open a file or directory in VS Code (uses 'code' CLI)."}, openFile)
	return server
}

func main() {
	// Flags to choose transport and address/path for HTTP mode
	httpMode := flag.Bool("http", false, "Serve over Streamable HTTP instead of stdio")
	addr := flag.String("addr", ":8081", "HTTP listen address (host:port)")
	mcpPath := flag.String("path", "/mcp", "HTTP path to mount the MCP handler")
	flag.Parse()

	if !*httpMode {
		// Default: stdio transport
		server := createServer()
		if err := server.Run(context.Background(), &mcp.StdioTransport{}); err != nil {
			log.Fatal(err)
		}
		return
	}

	// HTTP Streamable transport using StreamableHTTPHandler
	server := createServer()
	handler := mcp.NewStreamableHTTPHandler(func(r *http.Request) *mcp.Server { return server }, nil)

	mux := http.NewServeMux()
	// Health endpoints
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte("ok"))
	})
	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte("healthy"))
	})
	// Mount handler at both /path and /path/ to avoid redirects/edge cases
	p := *mcpPath
	if p == "" {
		p = "/mcp"
	}
	if !strings.HasPrefix(p, "/") {
		p = "/" + p
	}
	// Ensure a variant without trailing slash
	mux.Handle(p, handler)
	if !strings.HasSuffix(p, "/") {
		mux.Handle(p+"/", handler)
	}

	srv := &http.Server{Addr: *addr, Handler: mux}
	go func() {
		log.Printf("MCP streamable HTTP server listening at http://%s%s\n", strings.TrimPrefix(*addr, ":"), p)
		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			log.Fatalf("HTTP server error: %v", err)
		}
	}()

	// Graceful shutdown on SIGINT/SIGTERM
	stop := make(chan os.Signal, 1)
	signal.Notify(stop, syscall.SIGINT, syscall.SIGTERM)
	<-stop
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	_ = srv.Shutdown(ctx)
}

File: mcp-server/python3/mcp_server.py
import os
import shlex
import asyncio
import logging
import contextlib
import inspect
from typing import List
from collections.abc import AsyncIterator

from mcp import types
from mcp.server import Server
from mcp.server.streamable_http_manager import StreamableHTTPSessionManager

from starlette.applications import Starlette
from starlette.routing import Mount
from starlette.types import Scope, Receive, Send

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("vscode-file-finder-mcp")

# MCP Server instance
mcp_server = Server("vscode-file-finder")

BIN_PATH = "./vscode-helper"

def _build_cmd(base: List[str]) -> List[str]:
    if not os.path.exists(BIN_PATH):
        raise FileNotFoundError(
            f"Missing helper binary at {BIN_PATH}. Build it first: go build -o vscode-helper"
        )
    return [BIN_PATH, *base]

async def _run_cmd(cmd: List[str]) -> str:
    logger.debug("Running command: %s", " ".join(shlex.quote(c) for c in cmd))
    proc = await asyncio.create_subprocess_exec(
        *cmd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout, stderr = await proc.communicate()
    if proc.returncode != 0:
        err = (stderr.decode() or stdout.decode() or f"exit code {proc.returncode}").strip()
        raise RuntimeError(err)
    return stdout.decode().strip()

async def _search_files_impl(
    name: str = "",
    content: str = "",
    directory: str = ".",
) -> List[types.TextContent]:
    """Internal implementation for search_files tool returning unstructured text blocks."""
    cmd = _build_cmd(["search"])
    if name:
        cmd += ["--name", name]
    if content:
        cmd += ["--content", content]
    if directory and directory != ".":
        cmd += ["--dir", directory]
    try:
        output = await _run_cmd(cmd)
        output = output or "(no matches)"
        return [types.TextContent(type="text", text=output)]
    except Exception as e:
        return [types.TextContent(type="text", text=f"Error searching: {e}")]

async def _open_file_impl(
    path: str,
    open_dir: bool = False,
) -> List[types.TextContent]:
    """Internal implementation for open_file tool returning unstructured text blocks."""
    if not path:
        return [types.TextContent(type="text", text="Error: 'path' is required")]
    cmd = _build_cmd(["open"])
    if open_dir:
        cmd.append("--dir")
    cmd.append(path)
    try:
        output = await _run_cmd(cmd)
        output = output or f"Opened: {path}"
        return [types.TextContent(type="text", text=output)]
    except Exception as e:
        return [types.TextContent(type="text", text=f"Error opening: {e}")]

# Tool definitions (JSON Schemas) used for list_tools response
_TOOL_DEFINITIONS: list[types.Tool] = [
    types.Tool(
        name="search_files",
        description="Search files by name and/or content starting at a directory.",
        inputSchema={
            "type": "object",
            "properties": {
                "name": {"type": "string", "description": "Glob or pattern for file names"},
                "content": {"type": "string", "description": "Substring / text to search inside files"},
                "directory": {"type": "string", "description": "Root directory to start search", "default": "."},
            },
            "required": [],
            "additionalProperties": False,
        },
        outputSchema=None,
    ),
    types.Tool(
        name="open_file",
        description="Open a file or directory in VS Code (uses helper binary).",
        inputSchema={
            "type": "object",
            "properties": {
                "path": {"type": "string", "description": "Path to file or directory"},
                "open_dir": {"type": "boolean", "description": "Treat path as directory", "default": False},
            },
            "required": ["path"],
            "additionalProperties": False,
        },
        outputSchema=None,
    ),
]

@mcp_server.list_tools()
async def _list_tools() -> list[types.Tool]:  # noqa: D401
    """Provide tool definitions to MCP clients."""
    return _TOOL_DEFINITIONS

@mcp_server.call_tool()
async def _call_tool(tool_name: str, arguments: dict) -> List[types.TextContent]:  # noqa: D401
    """Dispatch a tool call based on name -> implementation."""
    arguments = arguments or {}
    if tool_name == "search_files":
        return await _search_files_impl(
            name=arguments.get("name", ""),
            content=arguments.get("content", ""),
            directory=arguments.get("directory", "."),
        )
    if tool_name == "open_file":
        return await _open_file_impl(
            path=arguments.get("path", ""),
            open_dir=bool(arguments.get("open_dir", False)),
        )
    return [types.TextContent(type="text", text=f"Error: unknown tool '{tool_name}'")]

# Streamable HTTPSessionManager expects the MCP Server as its 'app'
session_manager = StreamableHTTPSessionManager(app=mcp_server)

async def mcp_endpoint(scope: Scope, receive: Receive, send: Send) -> None:
    # Delegate each HTTP request under /mcp/ to the session manager
    await session_manager.handle_request(scope, receive, send)

@contextlib.asynccontextmanager
async def lifespan(_app: Starlette) -> AsyncIterator[None]:
    # Initialize streaming task group via run() which returns an async context manager
    run_fn = getattr(session_manager, "run", None)
    if not callable(run_fn):
        logger.warning("Session manager has no run(); continuing without background tasks.")
        yield
        return

    run_cm = run_fn()
    # Detect context manager
    if hasattr(run_cm, "__aenter__") and hasattr(run_cm, "__aexit__"):
        logger.info("Starting MCP session manager run() context")
        await run_cm.__aenter__()
    else:
        # Fallback: maybe awaitable
        if inspect.isawaitable(run_cm):
            await run_cm
        else:
            logger.warning("run() returned unsupported object (%r); skipping init.", run_cm)

    try:
        logger.info("MCP session manager started")
        yield
    finally:
        if hasattr(run_cm, "__aexit__"):
            try:
                await run_cm.__aexit__(None, None, None)
            except Exception as e:
                logger.warning("Error exiting run() context: %s", e)
        else:
            # Try graceful shutdown methods if context not used
            for attr in ("close", "shutdown", "stop"):
                fn = getattr(session_manager, attr, None)
                if callable(fn):
                    try:
                        res = fn()
                        if inspect.isawaitable(res):
                            await res
                    except Exception as e:
                        logger.warning("Error during %s(): %s", attr, e)
                    break
        logger.info("MCP session manager stopped")

# Starlette app just routes /mcp/ to the session manager
starlette_app = Starlette(
    debug=False,
    routes=[
        # Mount with trailing slash to avoid 307 redirect for client config using /mcp/
        Mount("/mcp/", app=mcp_endpoint),
    ],
    lifespan=lifespan,
)

def _registered_tool_names() -> list[str]:
    return [t.name for t in _TOOL_DEFINITIONS]

def main() -> None:
    if not os.path.exists(BIN_PATH):
        logger.warning("Helper binary not found at %s. Build it: go build -o vscode-helper", BIN_PATH)
    logger.info("Starting streamable HTTP MCP server at http://127.0.0.1:8080/mcp/")
    logger.info("Registered tools: %s", ", ".join(_registered_tool_names()))
    import uvicorn
    uvicorn.run(starlette_app, host="127.0.0.1", port=8080, log_level="info")

if __name__ == "__main__":
    main()